pkg/kapp/clusterapply/converged_resource.go

```diff
@@ -41,6 +70,35 @@ func (c ConvergedResource) IsDoneApplying() (ctlresm.DoneApplyState, []string, e
 		return ctlresm.DoneApplyState{Done: true}, descMsgs, err
 	}
 
+	// Custom wait rule
+	wr := c.waitingRule
+	if wr.SupportsObservedGeneration || len(wr.FailureConditions) > 0 || len(wr.SuccessfulConditions) > 0 {
```

Comment:
once this is moved into separate object in resourcemisc (see other objects in that package), we should not need this conditional.